kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DomainMetadataSuite.scala

```diff
@@ -0,0 +1,475 @@
+import io.delta.kernel.utils.CloseableIterable.{emptyIterable, inMemoryIterable}
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.delta.DeltaLog
+import org.apache.spark.sql.delta.actions.{DomainMetadata => SparkDomainMetadata}
+import org.apache.spark.sql.delta.test.DeltaTestImplicits.OptimisticTxnTestHelper
+
+import java.util.Collections
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Seq
+
+class DomainMetadataSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBase {
+
+  private def assertDomainMetadata(
+      snapshot: SnapshotImpl,
+      expectedValue: Map[String, DomainMetadata]): Unit = {
+    assert(expectedValue === snapshot.getDomainMetadataMap.asScala)
+  }
+
+  private def assertDomainMetadata(
+      table: Table,
+      engine: Engine,
+      expectedValue: Map[String, DomainMetadata]): Unit = {
+    // Get the latest snapshot of the table
+    val snapshot = table.getLatestSnapshot(engine).asInstanceOf[SnapshotImpl]
+    assertDomainMetadata(snapshot, expectedValue)
+  }
+
+  private def createTxnWithDomainMetadatas(
+      engine: Engine,
+      tablePath: String,
+      domainMetadatas: Seq[DomainMetadata]): Transaction = {
+
+    val txnBuilder = createWriteTxnBuilder(TableImpl.forPath(engine, tablePath))
+      .asInstanceOf[TransactionBuilderImpl]
+
+    val txn = txnBuilder.build(engine).asInstanceOf[TransactionImpl]
+    txn.addDomainMetadatas(domainMetadatas.asJava)
+    txn
+  }
+
+  private def commitDomainMetadataAndVerify(
+      engine: Engine,
+      tablePath: String,
+      domainMetadatas: Seq[DomainMetadata],
+      expectedValue: Map[String, DomainMetadata]): Unit = {
+    // Create the transaction with domain metadata and commit
+    val txn = createTxnWithDomainMetadatas(engine, tablePath, domainMetadatas)
+    txn.commit(engine, emptyIterable())
+
+    // Verify the final state includes the expected domain metadata
+    val table = Table.forPath(engine, tablePath)
+    assertDomainMetadata(table, engine, expectedValue)
+  }
+
+  private def setDomainMetadataSupport(engine: Engine, tablePath: String): Unit = {
+    val protocol = new Protocol(
+      1, // minReaderVersion
+      7, // minWriterVersion
+      Collections.emptyList(), // readerFeatures
+      Seq("domainMetadata").asJava // writerFeatures
+    )
+
+    val protocolAction = SingleAction.createProtocolSingleAction(protocol.toRow)
+    val txn = createTxn(engine, tablePath, isNewTable = false, testSchema, Seq.empty)
+    txn.commit(engine, inMemoryIterable(toCloseableIterator(Seq(protocolAction).asJava.iterator())))
+  }
+
+  private def createTableWithDomainMetadataSupported(engine: Engine, tablePath: String): Unit = {
+    // Create an empty table
+    createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)
+      .commit(engine, emptyIterable())
+
+    // Set writer version and writer feature to support domain metadata
+    setDomainMetadataSupport(engine, tablePath)
+  }
```

Comment:
This is fine for now but at some point we need to consider how we want to enable setting table protocols/features cc @vkorukanti @scottsand-db